clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -156,6 +397,112 @@ else if (key.defaultValue == NO_DEFAULT_VALUE)
+     * the current configuration values.
+     */
+    public List<ConfigValue> validate(Map<String, String> connectorConfigs) {
+        List<String> configsWithNoParent = getConfigsWithNoParent();
+        Map<String, ConfigValue> configValues = new HashMap<>();
+
+        for (String name: configKeys.keySet()) {
+            configValues.put(name, new ConfigValue(name));
+        }
+
+        for (String name: configsWithNoParent) {
+            validate(name, connectorConfigs, configValues);
+        }
+        return new LinkedList<>(configValues.values());
+    }
+
+    private void ensureDependentsDefined() {
+        for (String configName: configKeys.keySet()) {
+            ConfigKey configKey = configKeys.get(configName);
+            List<String> dependents = configKey.dependents;
+            for (String dependent: dependents) {
+                if (!configKeys.containsKey(dependent)) {
+                    throw new ConfigException(dependent + " is  an dependent of " + configName + ". However, it is not defined.");
+                }
+            }
+        }
+    }
+
+    private List<String> getConfigsWithNoParent() {
+        Set<String> configs = configKeys.keySet();
+        Set<String> configsWithParent = new HashSet<>();
+
+        for (ConfigKey configKey: configKeys.values()) {
+            List<String> dependents = configKey.dependents;
+            configsWithParent.addAll(dependents);
+        }
+
+        List<String> configsWithNoParent = new LinkedList<>();
+        for(String config: configs) {
+            if (!configsWithParent.contains(config)) {
+                configsWithNoParent.add(config);
+            }
+        }
+        return configsWithNoParent;
+    }
+
+    private void validate(String name, Map<String, String> connectorConfigs, Map<String, ConfigValue> configs) {
+        ConfigKey key = configKeys.get(name);
+        ConfigValue config = configs.get(name);
+
+        Object value;
+        if (connectorConfigs.containsKey(key.name)) {
+            value = parseType(key.name, connectorConfigs.get(key.name), key.type);
+        } else if (key.defaultValue == NO_DEFAULT_VALUE) {
+            config.addErrorMessage("Missing required configuration \"" + key.name + "\" which has no default value.");
+            value = null;
+        } else {
+            value = key.defaultValue;
+        }
+
+        if (key.validator != null) {
+            try {
+                key.validator.ensureValid(key.name, value);
+            } catch (ConfigException e) {
+                config.addErrorMessage(e.getMessage());
+            }
+        }
+        config.value(value);
+
+        List<Object> recommendedValues;
+        if (key.recommender != null) {
+            try {
+                recommendedValues = key.recommender.validValues(name, connectorConfigs);
+                List<Object> originalRecommendedValues = config.recommendedValues();
+
+                if (!originalRecommendedValues.isEmpty()) {
```

Comment:
Do we really want to do this? I get that the intent is to restrict recommendations to the original set. But couldn't that set depend on some other setting, where the original set and current set don't overlap at all? In that case wouldn't restricting the choices to the original set be incorrect?